src/platform/qt/library/LibraryModel.cpp

```diff
@@ -0,0 +1,417 @@
+		for (const LibraryEntry* item : newItems) {
+			addEntryInternal(*item);
+		}
+		endInsertRows();
+	}
+}
+
+void LibraryModel::updateEntries(const QList<LibraryEntry>& items) {
+	QHash<QModelIndex, SpanSet> updatedSpans;
+	for (const LibraryEntry& item : items) {
+		QModelIndex idx = index(item.fullpath);
+		Q_ASSERT(idx.isValid());
+		int pos = m_gameIndex.value(item.fullpath, -1);
+		Q_ASSERT(pos >= 0);
+		*m_games[pos] = item;
+		updatedSpans[idx.parent()].add(pos);
+	}
+	for (auto iter = updatedSpans.begin(); iter != updatedSpans.end(); iter++) {
+		QModelIndex parent = iter.key();
+		SpanSet spans = iter.value();
+		spans.merge();
+		for (const SpanSet::Span& span : spans.spans) {
+			QModelIndex topLeft = index(span.left, 0, parent);
+			QModelIndex bottomRight = index(span.right, MAX_COLUMN, parent);
+			emit dataChanged(topLeft, bottomRight);
+		}
+	}
+}
+
+void LibraryModel::removeEntries(const QList<QString>& items) {
+	SpanSet removedRootSpans;
+	QHash<QString, SpanSet> removedTreeSpans;
+	int firstModifiedIndex = m_games.size();
+	for (const QString& item : items) {
+		int pos = m_gameIndex.value(item, -1);
+		Q_ASSERT(pos >= 0);
+		if (pos < firstModifiedIndex) {
+			firstModifiedIndex = pos;
+		}
+		LibraryEntry* entry = m_games[pos].get();
+		QModelIndex parent = indexForPath(entry->base);
+		Q_ASSERT(!m_treeMode || parent.isValid());
+		QList<const LibraryEntry*>& pathItems = m_pathIndex[entry->base];
+		int pathPos = pathItems.indexOf(entry);
+		Q_ASSERT(pathPos >= 0);
+		removedTreeSpans[entry->base].add(pathPos);
+		if (!m_treeMode) {
+			removedRootSpans.add(pos);
+		}
+		m_gameIndex.remove(item);
+	}
+	for (const QString& base : removedTreeSpans.keys()) {
+		SpanSet& spanSet = removedTreeSpans[base];
+		spanSet.merge();
+		QList<const LibraryEntry*>& pathIndex = m_pathIndex[base];
+		if (spanSet.spans.size() == 1) {
+			SpanSet::Span span = spanSet.spans[0];
+			if (span.left == 0 && span.right == pathIndex.size() - 1) {
+				if (m_treeMode) {
+					removedRootSpans.add(m_pathOrder.indexOf(base));
+				} else {
+					m_pathIndex.remove(base);
+					m_pathOrder.removeAll(base);
+				}
+				continue;
+			}
+		}
+		QModelIndex parent = indexForPath(base);
+		spanSet.sort(true);
+		for (const SpanSet::Span& span : spanSet.spans) {
+			if (m_treeMode) {
+				beginRemoveRows(parent, span.left, span.right);
+			}
+			pathIndex.erase(pathIndex.begin() + span.left, pathIndex.begin() + span.right + 1);
+			if (m_treeMode) {
+				endRemoveRows();
+			}
+		}
+	}
+	removedRootSpans.merge();
+	removedRootSpans.sort(true);
+	for (const SpanSet::Span& span : removedRootSpans.spans) {
+		beginRemoveRows(QModelIndex(), span.left, span.right);
+		if (m_treeMode) {
+			for (int i = span.right; i >= span.left; i--) {
+				QString base = m_pathOrder.takeAt(i);
+				m_pathIndex.remove(base);
+			}
+		} else {
+			m_games.erase(m_games.begin() + span.left, m_games.begin() + span.right + 1);
```

Comment:
Hmm... I think I need to add another case to the unit test, because I'm not sure it ever frees the memory in tree mode...

... _oh._ THAT'S what that empty loop I removed was for. I wonder how the loop contents got removed.